### PR TITLE
Workpaths

### LINKED ISSUE #5: Backend implementation: non-linear Content work path
## Description
Implement the non-linear content work path in the backend according to the concept

## Acceptance criteria

- [ ] non-linear paths exist in the backend

### PR DESCRIPTION
**Pull request by BlueWaves2**
## Description of changes
Implementation of Work-Paths for Content.
Added:
- Work-Paths
- Stages (part of a work Path)

Closes #5 

## How has this been tested?
- [X] automated unit test
- [X] automated integration test
- [X] automated acceptance test
- [X] manual, exploratory test
    
## Checklist before requesting a review
- [X] My code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My code fulfils all acceptance criteria
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] The test coverage (line and branch) is reasonably high, especially on Service classes and other classes with complex logic
- [X] I have made corresponding changes to the documentation
- [X] I have added explanation of architectural decision and rationales to [wiki/adr](https://github.com/IT-REX-Platform/wiki/tree/main/adr)
- [X] I have updated the changes in the ticket description

## Checklist for reviewer
- [ ] The code works and does not throw errors
- [ ] The code is easy to understand and there are no confusing parts
- [ ] The code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] The code change accomplishes what it is supposed to do
- [ ] I cannot think of any use case in which the code does not behave as intended
- [ ] The added and existing tests reasonably cover the code change
- [ ] I cannot think of any test cases, input or edge cases that should be tested in addition
- [ ] Description of the change is included in the documentation

### CODE REVIEW

File: src/main/resources/graphql/service/mutation.graphqls [hunk 1]
```diff
@@ -31,6 +31,41 @@ type Mutation {
...
+    """
+    createStage(workPathId: UUID!): Stage!
+
+    """
+    Update Content of Stage
+    """
+    updateStage(input: UpdateStageInput): Stage!
+
+    """
+    delete STage by ID
```
Comment (PaulBredl) [inline src/main/resources/graphql/service/mutation.graphqls:61]:
```suggestion
    delete Stage by ID
```

**Review by PaulBredl — COMMENTED** [review]
I just quickly looked over it, the review should ideally make @myluki2000, because he did the concept. 
From only a short look it looked good. I personally would have probably designed the API a bit differently, but as long as the frontend is happy with it, I am too :D

File: src/main/resources/graphql/service/workPath.graphqls [hunk 2]
```diff
@@ -0,0 +1,13 @@
+type WorkPath {
```
Comment (myluki2000) [inline src/main/resources/graphql/service/workPath.graphqls:1]: Please add doc-comments to all fields in these 2 types

File: src/main/resources/graphql/service/workPath.graphqls [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+    id: UUID!
+    name: String!
+    chapterId: UUID!
+    stages: [Stage!]!
+}
+
+type Stage{
+    id: UUID!
+    position: Int!
+    requiredContents: [Content!]!
```
Comment (myluki2000) [inline src/main/resources/graphql/service/workPath.graphqls:11]: I would've liked a more generic solution instead of just storing a list of required and a list of optional contents, but @PaulBredl would probably hate me for it

Comment (myluki2000) [reply src/main/resources/graphql/service/workPath.graphqls:11]: Don't we also need a way to store what counts as "completed"? E.g. setting a minimum correctness score for a quiz. Or are we storing that in the quiz service?

**Review by myluki2000 — CHANGES_REQUESTED** [review]

File: src/main/resources/graphql/service/workPath.graphqls (hunk 3 above)
Comment (BlueWaves2) [reply src/main/resources/graphql/service/workPath.graphqls:11]: You should be able to derive completeness from the content. 
Note, current implementation only contains necessary data to create and modify Work Paths & Stages.
We will add User specific data later in following sprints

File: src/main/resources/graphql/service/workPath.graphqls (hunk 2 above)
Comment (BlueWaves2) [reply src/main/resources/graphql/service/workPath.graphqls:4]: all graphql representations are now commented

File: src/main/resources/graphql/service/workPath.graphqls (hunk 3 above)
Comment (myluki2000) [reply src/main/resources/graphql/service/workPath.graphqls:45]: As far as I can remember our consensus was that the definition of completeness should be settable in the stage and not the content itself (as content may be re-used in mulitple stages).

But even if not, a differentiation betwee requiredContent and optionalContent does not make sense in any case because if a content is optional you could just set its minimum correctness to 0.

@PaulBredl opinions?

Comment (PaulBredl) [reply src/main/resources/graphql/service/workPath.graphqls:45]: > I would've liked a more generic solution instead of just storing a list of required and a list of optional contents, but @PaulBredl would probably hate me for it

@myluki2000 What do mean with more generic solution? Isn't this implementation what we discussed on Monday?

Comment (PaulBredl) [reply src/main/resources/graphql/service/workPath.graphqls:45]: > Don't we also need a way to store what counts as "completed"? E.g. setting a minimum correctness score for a quiz. Or are we storing that in the quiz service?

Stored in quiz service

Comment (PaulBredl) [reply src/main/resources/graphql/service/workPath.graphqls:45]: > But even if not, a differentiation betwee requiredContent and optionalContent does not make sense in any case because if a content is optional you could just set its minimum correctness to 0.

That doesn't make sense to me? How does a media content like a pdf have a correctness?

Comment (myluki2000) [reply src/main/resources/graphql/service/workPath.graphqls:45]: > That doesn't make sense to me? How does a media content like a pdf have a correctness?

It doesn't but it has some way of deciding what counts as done. In case of a quiz it's correctness, in case of a PDF it's if it has been read or not

Comment (myluki2000) [reply src/main/resources/graphql/service/workPath.graphqls:45]: > @myluki2000 What do mean with more generic solution? Isn't this implementation what we discussed on Monday?

Didn't we say on monday that the lecturer should be able to set in the stage what counts as completed or not?

In any case, having BOTH a minimum correctness in the quiz service and an "optional content" here in the content service is a duplication to me, they both map to the same concept and should not both exist at the same time. I.e. why should an optional quiz have a "minimum passing grade" when you don't need to pass it in the first place?

Comment (PaulBredl) [reply src/main/resources/graphql/service/workPath.graphqls:45]: > In any case, having BOTH a minimum correctness in the quiz service and an "optional content" here in the content service is a duplication to me, they both map to the same concept and should not both exist at the same time. I.e. why should an optional quiz have a "minimum passing grade" when you don't need to pass it in the first place?

No that is not a duplication at all. It actually makes more complex things possible, like saying you need 50% in this quiz and 70% in the other to complete the course.
Also you cannot make a quiz optional by making the requiredCorrectness 0. That doesn't make any sense? Even in an optional quiz you want to have feedback if you passed the quiz or not. If you just set it to 0 means:
1) in order the complete it you would still have to click through it once (akward)
2) even if I click everything wrong, the feedback would still be "Yay you completed the quiz, you're amazing" and the learning interval will be higher, etc.

Comment (PaulBredl) [reply src/main/resources/graphql/service/workPath.graphqls:45]: > Didn't we say on monday that the lecturer should be able to set in the stage what counts as completed or not so the student can progress to the next stage?

Yes by marking which contents are required and which are optional

**Review by myluki2000 — APPROVED** [review]

**Review by PaulBredl — APPROVED** [review]
